attestationreport/sgx.go

```diff
@@ -0,0 +1,367 @@
+		return nil, true
+	}
+
+	if len(referenceValues) == 0 {
+		msg := "Could not find SGX Reference Value"
+		result.Summary.setFalse(&msg)
+		return result, false
+	} else if len(referenceValues) > 1 {
+		msg := fmt.Sprintf("Report contains %v reference values. Currently, only one SGX Reference Value is supported",
+			len(referenceValues))
+		result.Summary.setFalse(&msg)
+		return result, false
+	}
+	sgxReferenceValue := referenceValues[0]
+
+	// If the attestationreport contains SGX Reference Values, but no SGX measurement, the attestation must fail
+	if sgxM == nil {
+		for _, v := range referenceValues {
+			result.Artifacts = append(result.Artifacts,
+				DigestResult{
+					Name:    v.Name,
+					Digest:  hex.EncodeToString(v.Sha256),
+					Success: false,
+					Type:    "Reference Value",
+				})
+		}
+		result.Summary.Success = false
+		return result, false
+	}
+
+	// Validate Parameters:
+	if sgxM.Report == nil || len(sgxM.Report) < SGX_QUOTE_MIN_SIZE {
+		msg := "Invalid SGX Report."
+		result.Summary.setFalse(&msg)
+		return result, false
+	}
+
+	if sgxReferenceValue.Type != "SGX Reference Value" {
+		msg := fmt.Sprintf("SGX Reference Value invalid type %v", sgxReferenceValue.Type)
+		result.Summary.setFalse(&msg)
+		return result, false
+	}
+
+	if sgxReferenceValue.Sgx == nil {
+		msg := "SGX Reference Value is null"
+		result.Summary.setFalse(&msg)
+		return result, false
+	}
+
+	var sgxQuote SgxReport
+	var quoteType uint32 = sgxReferenceValue.Sgx.Collateral.TeeType
+	if quoteType == SGX_QUOTE_TYPE {
+		// extract the attestation report into the SGXReport data structure
+		sgxQuote, err = DecodeSgxReport(sgxM.Report)
+		if err != nil {
+			msg := fmt.Sprintf("Failed to decode SGX report: %v", err)
+			result.Summary.setFalse(&msg)
+			return result, false
+		}
+	} else {
+		log.Tracef("Unknown quote type (tee_type: %X)\n", quoteType)
+		return result, false
+	}
+
+	// Compare Nonce for Freshness (called Report Data in the SNP Attestation Report Structure)
+	// ReportData contains: nonce in ReportData field
+	nonce64 := make([]byte, 64)
+	copy(nonce64, nonce[:])
+
+	if cmp := bytes.Compare(sgxQuote.ISVEnclaveReport.ReportData[:], nonce64); cmp != 0 {
+		msg := fmt.Sprintf("Nonces mismatch: Plain Nonce: %v, Expected: %v, Got = %v",
+			nonce, hex.EncodeToString(nonce64), hex.EncodeToString(sgxQuote.ISVEnclaveReport.ReportData[:]))
+		result.Freshness.setFalse(&msg)
+		result.Summary.Success = false
+		return result, false
+	} else {
+		result.Freshness.Success = true
+	}
+
+	// parse cert chain
+	referenceCerts, err := parseCertificates(sgxM.Certs, true)
+	if err != nil {
+		msg := fmt.Sprintf("Failed to parse reference certificates: %v", err)
+		result.Summary.setFalse(&msg)
+		return result, false
+	}
+
+	// (from DCAP Library): parse PCK Cert chain (from the quote) into CertificateChain object. return error in case of failure
+	// TODO: handle other QECertDataTypes (for now: throw an error)
```

Comment:
hältst du es für sinnvoll/realistisch, dass wir zeitnah auch andere QECertDataTypes unterstützen?
Falls nicht, würde ich vllt. statt der etwas provokanten "TODO"-Formulierung eher sagen "Currently only ... is supported"